### PR TITLE
Material package fixes

### PR DESCRIPTION
**Pull request by mic4ael**
* Filtering attachments by contribution schedule dates
* Generating proper folder structure for generated ZIP file

### CODE REVIEW

File: indico/modules/attachments/controllers/event_package.py [hunk 1]
```diff
@@ -102,14 +117,27 @@ def _generate_zip_file(self, attachments):
...
+            start_date = linked_object.parent.getAdjustedStartDate()
+            path_parts.extend([start_date.strftime('%Y%m%d_%A'),
+                               start_date.strftime('%H%M_{}').format(linked_object.parent.getTitle()),
+                               linked_object.getTitle()])
+        elif linked_object != self._conf:
+            start_date = linked_object.getAdjustedStartDate()
+            path_parts.extend([start_date.strftime('%Y%m%d_%A'),
+                               start_date.strftime('%H%M_{}').format(linked_object.getTitle())])
+
+        path_parts.extend([attachment.folder.title or '', attachment.file.filename])
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: secure_filename(attachment.folder.title, unicode(attachment.folder.id))

File: indico/modules/attachments/controllers/event_package.py [hunk 2]
```diff
@@ -102,14 +117,27 @@ def _generate_zip_file(self, attachments):
...
-
+                name = self._prepare_folder_structure(attachment)
                 with attachment.file.storage.get_local_path(attachment.file.storage_file_id) as filepath:
                     zip_handler.write(filepath, name)
 
         return send_file('material-{}.zip'.format(self._conf.id), temp_file.name, 'application/zip', inline=False)
+
+    def _prepare_folder_structure(self, attachment):
+        linked_object = attachment.folder.linked_object
+        path_parts = [self._conf.getTitle()]
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: secure_filename(self._conf.getTitle())

File: indico/modules/attachments/controllers/event_package.py [hunk 3]
```diff
@@ -102,14 +117,27 @@ def _generate_zip_file(self, attachments):
...
+
+        if isinstance(linked_object, SubContribution):
+            start_date = linked_object.parent.getAdjustedStartDate()
+            path_parts.extend([start_date.strftime('%Y%m%d_%A'),
+                               start_date.strftime('%H%M_{}').format(linked_object.parent.getTitle()),
+                               linked_object.getTitle()])
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: secure_filename(linked_object.getTitle(), unicode(linked_object.getId()))

File: indico/modules/attachments/controllers/event_package.py [hunk 4]
```diff
@@ -102,14 +117,27 @@ def _generate_zip_file(self, attachments):
...
         return send_file('material-{}.zip'.format(self._conf.id), temp_file.name, 'application/zip', inline=False)
+
+    def _prepare_folder_structure(self, attachment):
+        linked_object = attachment.folder.linked_object
+        path_parts = [self._conf.getTitle()]
+
+        if isinstance(linked_object, SubContribution):
+            start_date = linked_object.parent.getAdjustedStartDate()
+            path_parts.extend([start_date.strftime('%Y%m%d_%A'),
+                               start_date.strftime('%H%M_{}').format(linked_object.parent.getTitle()),
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: secure_filename(start_date.strftime('%H%M_{}').format(linked_object.parent.getTitle(), '')

no need for a good default here; with the prefix you'll always have something valid

**Comment by ThiefMaster** [comment]
https://github.com/ThiefMaster/indico/tree/attachment-package - rebased and I already fixed some small things I've noticed.

However, there is still a rather problematic issue in there: Right now "sessions", "contributions" and "contributions scheduled on" are cumulative. So you need to uncheck all the contributions if you e.g. want to get only a certain day. That's rather terrible UX.

My solution to this would be unchecking everything by default and only applying the filter when something is selected. So if nothing is selected at all, everything is included. As soon as you select something in a group, those elements are included. This actually lets you e.g. select a whole day plus some extra contributions/sessions.

**Comment by ThiefMaster** [comment]
Also, unscheduled contributions are still included in the contribution list... I still think they shouldn't be shown (@pferreir what do you think?)

**Comment by ThiefMaster** [comment]
Actually... I wonder if selecting a session should include only material from that session or from everything inside that session.

**Comment by pferreir** [comment]
I think unscheduled contribs should not be shown, indeed. Unless we add another option "include unscheduled contributions". That could be a solution.

**Comment by pferreir** [comment]
And I agree, by "session" I understand the whole thing, not just the session object itself. And from the user's perspective, I believe that is what is expected.

File: indico/modules/attachments/controllers/event_package.py [hunk 5]
```diff
@@ -80,15 +93,28 @@ def _build_base_query(self):
                                _join=AttachmentFolder)
 
     def _filter_by_sessions(self, session_ids, added_since):
-        query = self._build_base_query().filter(AttachmentFolder.session_id.in_(session_ids))
+        query = self._build_base_query()
+        if session_ids:
+            query = query.filter(AttachmentFolder.session_id.in_(session_ids))
+        else:
+            query = query.filter(AttachmentFolder.session_id != None)
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]:
```
query = query.filter(AttachmentFolder.link_type == LinkType.session)
if session_ids:
    query = query.filter(AttachmentFolder.session_id.in_(session_ids))
```

File: indico/modules/attachments/controllers/event_package.py [hunk 6]
```diff
@@ -99,17 +125,39 @@ def _filter_by_date(self, query, added_since):
...
-
+                name = self._prepare_folder_structure(attachment)
                 with attachment.file.storage.get_local_path(attachment.file.storage_file_id) as filepath:
                     zip_handler.write(filepath, name)
 
         return send_file('material-{}.zip'.format(self._conf.id), temp_file.name, 'application/zip', inline=False)
+
+    def _prepare_folder_structure(self, attachment):
+        event_dir = secure_filename(self._conf.getTitle(), None)
+        path_paths = [event_dir] if event_dir else []
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: `path_paths`?! `path_segments` or `path_parts` would be better fits.. or maybe even just `segments`

File: indico/modules/attachments/controllers/event_package.py [hunk 7]
```diff
@@ -99,17 +125,39 @@ def _filter_by_date(self, query, added_since):
...
+        path_paths.extend(self._get_base_path(attachment))
+        path_paths.extend([secure_filename(attachment.folder.title, unicode(attachment.folder.id)),
+                           attachment.file.filename])
+        return os.path.join(*path_paths)
+
+    def _get_base_path(self, attachment):
+        linked_object = attachment.folder.linked_object
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: I'd rename this to `obj`. later in the loop it's not the linked objecy anymore.

File: indico/modules/attachments/controllers/event_package.py [hunk 8]
```diff
@@ -99,17 +125,39 @@ def _filter_by_date(self, query, added_since):
...
+        paths = []
+        while linked_object != self._conf:
+            owner = linked_object.getOwner()
+            if isinstance(linked_object, SubContribution):
+                start_date = owner.getAdjustedStartDate()
+            else:
+                start_date = linked_object.getAdjustedStartDate()
+
+            paths.insert(0, secure_filename(start_date.strftime('%H%M_{}').format(linked_object.getTitle()), ''))
```
Comment (ThiefMaster) [inline indico/modules/attachments/controllers/event_package.py]: why not append and in the end `return reversed(paths)`?

File: indico/modules/attachments/forms.py [hunk 9]
```diff
@@ -104,3 +104,7 @@ class AttachmentPackageForm(IndicoForm):
                                                  description=_('Include attachments from selected sessions'))
     contributions = IndicoSelectMultipleCheckboxField(_('Contributions'),
                                                       description=_('Include attachments from selected contributions'))
+    contributions_schedule_dates = IndicoSelectMultipleCheckboxField(_('Contributions scheduled on'),
```
Comment (ThiefMaster) [inline indico/modules/attachments/forms.py]: "Dates" might be a better label for this. I guess it doesn't only apply to contributions but also e.g. sessions

File: indico/modules/attachments/controllers/event_package.py (hunk 6 above)
Comment (mic4ael) [inline indico/modules/attachments/controllers/event_package.py]: I don't know how..

File: indico/modules/attachments/controllers/event_package.py (hunk 8 above)
Comment (mic4ael) [inline indico/modules/attachments/controllers/event_package.py]: I thought about that. I must check which option is better.

File: indico/modules/attachments/forms.py (hunk 9 above)
Comment (mic4ael) [inline indico/modules/attachments/forms.py]: That is the same as in the previous class responsible for material packages.

File: indico/MaKaC/webinterface/rh/conferenceModif.py [hunk 10]
```diff
@@ -66,6 +66,7 @@
 from indico.util import json
 from indico.web.http_api.metadata.serializer import Serializer
 from indico.web.flask.util import send_file, url_for
+from indico.modules.attachments.controllers.event_package import AttachmentPackageGeneratorMixin
```
Comment (ThiefMaster) [inline indico/MaKaC/webinterface/rh/conferenceModif.py]: `m<u`

**Comment by ThiefMaster** [comment]
Fixed some issues: https://github.com/ThiefMaster/indico/commits/attachment-package

Unfortunately days are still not handled properly. With my [test event](https://blackhole.cern.ch/event/293836/) I get a ZIP file containing `Requests_test\20150902_Wednesday\1335_S` even though I selected only `03/09/2015`

**Comment by mic4ael** [comment]
The reason why it works like that is because - just like you proposed some time ago - in case no filters are being selected in a section (e.g sessions, contributions) then all of the attachments within it are included. So I think we should change the way filtering works right now.

**Pull request merged**